src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,365 @@
+            }
+        }
+
+        /// <summary>
+        /// Equivalent to <see cref="TryGetDelegate{TDelegate}(string, bool, out TDelegate)"/> where <em>exactSpelling</em> is <see langword="false"/>.
+        /// </summary>
+        public bool TryGetDelegate<TDelegate>(string name, out TDelegate result) where TDelegate : class
+            => TryGetDelegate<TDelegate>(name, exactSpelling: false, out result);
+
+        /// <summary>
+        /// Attempts to locate a named function in the module's export table and return a managed delegate to that function.
+        /// </summary>
+        /// <typeparam name="TDelegate">The delegate type that will be created. This type must subclass <see cref="MulticastDelegate"/>.
+        /// This type <em>should</em> be (but is not required to be) annotated with <see cref="UnmanagedFunctionPointerAttribute"/>.</typeparam>
+        /// <param name="name">The name of the function to look up.</param>
+        /// <param name="exactSpelling"><see langword="false"/> (default) if the runtime should automatically perform name decoration when
+        /// locating the exported function; <see langword="true"/> if the runtime should use the provided name as-is with no additional
+        /// decoration. See also <see cref="DllImportAttribute.ExactSpelling"/>.</param>
+        /// <param name="result">A <typeparamref name="TDelegate"/> that represents the exported function,
+        /// or <see langword="null"/> if the function cannot be located in the export table.</param>
+        /// <returns><see langword="true"/> if the function was found, otherwise <see langword="false"/>.</returns>
+        /// <remarks>
+        /// Functions can also be queried by ordinal export by using a "#" prefix (e.g., "#1") for <paramref name="name"/>.
+        /// </remarks>
+        public bool TryGetDelegate<TDelegate>(string name, bool exactSpelling, out TDelegate result) where TDelegate : class
+        {
+            // Parameter validation
+            // Some checks overlap with Marshal.GetDelegateForFunctionPointer since we need to
+            // call into the CLR separately so that we can support mangling and other features.
+
+            var delegateType = typeof(TDelegate) as RuntimeType;
+
+            // Must be a normal RuntimeType and the immediate parent class must be MulticastDelegate
+            if (delegateType == null || !delegateType.IsDelegate())
+            {
+                throw new ArgumentException(
+                    paramName: nameof(TDelegate),
+                    message: SR.Arg_MustBeDelegate);
+            }
+
+            RuntimeMethodHandleInternal invokeMethodHandle = new RuntimeMethodHandleInternal(StubHelpers.StubHelpers.GetDelegateInvokeMethodFromDelegateType(delegateType));
+            RuntimeMethodInfo invokeMethodInfo = (RuntimeMethodInfo)RuntimeType.GetMethodBase(delegateType, invokeMethodHandle);
+
+            // Generic types are disallowed
```

Comment:
why generic delegates is n't allowed? Func<T> or Action<T> should be okay.

Reply:
I assume that the check here was supposed to be for `delegateType.ContainsGenericParameters`, aka `delegateType.IsOpenType` (that property doesn't actually exist).

Using `ContainsGenericParameters` allows `typeof(Func<bool>)`, but not `typeof(Func<>)`.

Reply:
It's an existing limitation of [Marshal.GetDelegateForFunctionPointer](https://msdn.microsoft.com/en-us/library/zdx6dyyh(v=vs.110).aspx). Many of these checks are duplicates of the checks performed by that method; I'm just performing the checks upfront so that we don't go too far down the "find the entry point" logic if we know we're going to fail immediately after.

Reply:
Huh.  Weird.  Well, that's sufficient for me.  (Though now I wonder why GetDelegateForFunctionPointer has that restriction.  But, it already exists, so OK.)

Reply:
If I had to guess (and this is only a guess), I'd venture that the internal marshaling system has trouble specializing generics for reference types. For example, consider `Func<string, int>` and `Func<SafeHandle, int>`. They're both represented under the covers as a single `Func<__Canon, int>` type, but `string` and `SafeHandle` have very different requirements for marshaling. I suppose it would be possible to support this but due to lack of interest nobody ever took the time to allow it.